src/rt-app_parse_config.c

```diff
@@ -838,6 +904,15 @@ parse_thread_data(char *name, struct json_object *obj, int index,
 		assure_type_is(phases_obj, obj, "phases",
 					json_type_object);
 
+		/*
+		 * Taskgroup must not be specified at thread level if phases
+		 * are present.
+		 */
```

Comment:
Sorry, I haven't read all previous comment but why do we add such constraint on TG whereas we don't have such for other events ?

Reply:
policy/event difference - @deggeman @vingu-linaro The taskgroup is written as a policy rather than an event, which I also didn't notice at first. As a policy, these restrictions are unavoidable.

Reply:
I would like to handle a taskgroup like a cpuset or sched_data (sched class and priority). To me it makes more sense this way. A task or a phase of a task runs in a certain taskgroup like with a certain CPU affinity or in a certain sched class with a certain priority. Events are even more dynamic and they often work in pairs. I don't see why switching to a certain taskgroup has to be more fine grained than the phase border.

Reply:
:+1:

Reply:
cpuset and sched_data have a different behavior and I prefer the sched_data behavior more than cpuset one because it enables incremental/decremental changes in the json file.
The fact that p1 in your example above doesn't have the same policy as p2 and following one is not a problem IMO. In your example, the problem comes from the json file

Reply:
This can be easily changed by reverting patch ("taskgroups: Thread/phase taskgroup config check") of this set:

```diff --git a/src/rt-app_parse_config.c b/src/rt-app_parse_config.c
index b286589e2a77..a19297d42266 100644
--- a/src/rt-app_parse_config.c
+++ b/src/rt-app_parse_config.c
@@ -1008,15 +1008,6 @@ parse_task_data(char *name, struct json_object *obj, int index,
                assure_type_is(phases_obj, obj, "phases",
                                        json_type_object);
 
-               /*
-                * Taskgroup must not be specified at thread level if phases
-                * are present.
-                */
-               if (strlen(get_string_value_from(obj, "taskgroup", TRUE, ""))) {
-                       log_critical("Specify taskgroup in phase rather than thread");
-                       exit(EXIT_INV_CONFIG);
-               }
-
                log_info(PIN "Parsing phases section");
                data->nphases = 0;
                foreach(phases_obj, entry, key, val, idx) {```

Letting the first phase run in different taskgroups is not an error, it could just confuse people but it can be easily changed in the json file in case it is an issue.

Reply:
Testing has revealed that I missed setting the taskgroup for the thread. So the following snippets have to be added:

```
    Fix 1: for 'taskgroups: Data structures and bricks for config parser'
    
    Signed-off-by: Dietmar Eggemann <dietmar.eggemann@arm.com>

diff --git a/src/rt-app_types.h b/src/rt-app_types.h
index 8025a56b004a..cf34a752ccd8 100644
--- a/src/rt-app_types.h
+++ b/src/rt-app_types.h
@@ -205,7 +205,8 @@ typedef struct _thread_data_t {
        sched_data_t *sched_data; /* scheduler policy information */
        sched_data_t *curr_sched_data; /* current scheduler policy */
 
-       taskgroup_data_t *curr_taskgroup_data;
+       taskgroup_data_t *taskgroup_data; /* taskgroup information */
+       taskgroup_data_t *curr_taskgroup_data; /* current taskgroup */
 
        int loop;
        int nphases;
```
```    Fix 2: for 'taskgroups: Config parser'
    
    Signed-off-by: Dietmar Eggemann <dietmar.eggemann@arm.com>

diff --git a/src/rt-app_parse_config.c b/src/rt-app_parse_config.c
index 282aa4b4e029..b36e45e1edaa 100644
--- a/src/rt-app_parse_config.c
+++ b/src/rt-app_parse_config.c
@@ -969,6 +969,8 @@ parse_task_data(char *name, struct json_object *obj, int index,
        parse_cpuset_data(obj, &data->cpu_data);
        /* Scheduling policy */
        data->sched_data = parse_sched_data(obj, opts->policy);
+       /* Taskgroup */
+       data->taskgroup_data = parse_taskgroup_data(obj);
 
        /* initial delay */
```
```    Fix 3: for 'taskgroups: Taskgroup thread setter'
    
    Signed-off-by: Dietmar Eggemann <dietmar.eggemann@arm.com>

diff --git a/src/rt-app.c b/src/rt-app.c
index 999353dc6600..fd7e90afd576 100644
--- a/src/rt-app.c
+++ b/src/rt-app.c
@@ -1068,6 +1068,7 @@ void *thread_body(void *arg)
                        data->ind, policy_to_string(data->sched_data->policy),
                        data->sched_data->prio);
        set_thread_priority(data, data->sched_data);
+       set_thread_taskgroup(data, data->taskgroup_data);
 
        /*
         * phase        - index of current phase in data->phases array
```
```
    Fix 4: for 'taskgroups: Policy/taskgroup dependency check'
    
    Signed-off-by: Dietmar Eggemann <dietmar.eggemann@arm.com>

diff --git a/src/rt-app_parse_config.c b/src/rt-app_parse_config.c
index 65cf3f88dc2d..aa82b022e504 100644
--- a/src/rt-app_parse_config.c
+++ b/src/rt-app_parse_config.c
@@ -996,7 +996,7 @@ parse_task_data(char *name, struct json_object *obj, int index,
         * policy/taskgroup misconfiguration.
         */
        data->curr_sched_data = data->sched_data;
-       data->curr_taskgroup_data = NULL;
+       data->curr_taskgroup_data = data->taskgroup_data;
 
        /* initial delay */
```